ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZones.ts

```diff
@@ -4,58 +4,202 @@ import {
+const Y_OFFSET_INCREMENT = 4;
+const MAX_ZONES = 9;
+const X_BACKGROUND_OFFSET = 22;
+const X_TROUGHTRAIN_PADDING = 4;
+const Y_BACKGROUND_OFFSET = 20;
+const SELECTED_TRAIN_ID_GRADIANT = 2;
+
+type DrawZone = {
+  ctx: CanvasRenderingContext2D;
+  arrivalTime: number;
+  departureTime: number;
+  yPosition: number;
+};
+
+const drawDefaultZone = ({ ctx, arrivalTime, departureTime, yPosition }: DrawZone) => {
+  ctx.beginPath();
+  ctx.rect(arrivalTime, yPosition, departureTime! - arrivalTime, OCCUPANCY_ZONE_HEIGHT);
+  ctx.fill();
+  ctx.stroke();
+};
+
+const ARROW_OFFSET_X = 1;
+const ARROW_OFFSET_Y = 1.5;
+const ARROW_WIDTH = 4.5;
+const ARROW_TOP_Y = 3.5;
+const ARROW_BOTTOM_Y = 6.5;
+
+const drawThroughTrain = ({ ctx, arrivalTime }: Omit<DrawZone, 'departureTime' | 'yPosition'>) => {
+  ctx.beginPath();
+  ctx.moveTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime - ARROW_WIDTH, OCCUPANCY_ZONE_START - ARROW_TOP_Y);
+  ctx.lineTo(arrivalTime + ARROW_WIDTH, OCCUPANCY_ZONE_START - ARROW_TOP_Y);
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime + ARROW_WIDTH, OCCUPANCY_ZONE_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTime - ARROW_WIDTH, OCCUPANCY_ZONE_START + ARROW_BOTTOM_Y);
+  ctx.lineTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.fill();
+  ctx.moveTo(arrivalTime - ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.lineTo(arrivalTime + ARROW_OFFSET_X, OCCUPANCY_ZONE_START + ARROW_OFFSET_Y);
+  ctx.stroke();
+};
+
+type DrawRemainingTrainsBox = {
+  ctx: CanvasRenderingContext2D;
+  text: string;
+  textX: number;
+  textY: number;
+};
+
+const drawRemainingTrainsBox = ({ ctx, text, textX, textY }: DrawRemainingTrainsBox) => {
+  ctx.fillStyle = REMAINING_TRAINS_BACKGROUND;
+  ctx.beginPath();
+  ctx.rect(textX, textY, REMAINING_TRAINS_WIDTH, REMAINING_TRAINS_HEIGHT);
+  ctx.fill();
+  ctx.stroke();
+  ctx.fillStyle = WHITE_100;
+  ctx.font = SANS;
+  ctx.textAlign = 'center';
+  ctx.textBaseline = 'middle';
+  ctx.fillText(text, textX + REMAINING_TRAINS_WIDTH / 2, textY + REMAINING_TRAINS_HEIGHT / 2);
+};
+
 export const drawOccupancyZones = ({
   ctx,
   width,
   height,
   tracks,
   occupancyZones,
   getTimePixel,
+  selectedTrainId,
 }: {
   ctx: CanvasRenderingContext2D;
   width: number;
   height: number;
   tracks: Track[] | undefined;
   occupancyZones: OccupancyZone[] | undefined;
   getTimePixel: (time: number) => number;
+  selectedTrainId: string;
 }) => {
   ctx.clearRect(0, 0, width, height);
 
   tracks?.forEach((track, index) => {
     const trackTranslate = index === 0 ? CANVAS_PADDING : TRACK_HEIGHT_CONTAINER;
     ctx.translate(0, trackTranslate);
 
-    const trackOccupancyZones = occupancyZones?.filter((zone) => zone.trackId === track.id);
+    const filteredOccupancyZone = occupancyZones?.filter((zone) => zone.trackId === track.id);
```

Comment:
You can add an early return if there is no tracks or no occupancyZones, and remove all the `?`